docs/qa-metrics.md

```diff
@@ -0,0 +1,374 @@
+# Metrics for Software Quality Assurance
+
+- [Metrics for Software Quality Assurance](#metrics-for-software-quality-assurance)
+  - [Summary](#summary)
+  - [Code Quality Guidelines](#code-quality-guidelines)
+  - [Effectiveness of Code Quality Approaches](#effectiveness-of-code-quality-approaches)
+  - [Tools](#tools)
+  - [Best Practices](#best-practices)
+  - [Unsorted](#unsorted)
+
+## Summary
+
+This is a literature survey of metrics used to determine the quality of software.
+Each paper will have a summary of metrics, the correlation of metrics to software quality,
+definitions of quality, and a list of any available tools to generate the metrics.
+Note that not all papers will include each of those features.
+
+Some of the ways that quality is used:
+
+- Fault detection
+- Maintainability
+- Extensibility
+- Security
+
+## Code Quality Guidelines
+
+**CMU SEI CERT C++ Coding Standard** \
+2016 \
+<https://wiki.sei.cmu.edu/confluence/display/cplusplus/CC.+Analyzers>
+
+This is a set of rules and recommendations for writing secure code.
+The nice thing is that for each guideline the authors have identified static analysis tools and their
+respective flags that can be used to identify violations of the guidelines.
+There are examples showing violations and how to fix them.
+
+The open source tools they identify are: clang-tidy, cppcheck, findbugs, Make.
+Using these open source tools makes it straightforward to verify that a subset of the recommended standards
+are being applied to a particular code base.
+
+Statick has a specific configuration to use these open source tools with the flags recommended by this standard.
+Statick also extends the recommended tools to identify the same types of security issues for languages other than C++.
+
+**Software Quality Metrics for Object-Oriented Environments** \
+Dr. Linda H. Rosenberg and Lawrence E. Hyatt \
+Goddard Space Flight Center \
+<https://pdfs.semanticscholar.org/fc8c/16be91c43bb15b72e2c73728839c1a9468ef.pdf> \
+<https://people.ucalgary.ca/~far/Lectures/SENG421/PDF/oocross.pdf>
+
+The authors identify five attributes that contribute to quality software: efficiency, complexity,
+understandability, reusability, testability/maintainability.
+
+They then identify ways to measure features of software and tie those features back to the software quality attributes.
+The features they use are: cyclomatic complexity, size, comment percentage, weighted methods per class,
+response for a class (similar to afferent coupling), lack of cohesion of methods, coupling between object
+classes, depth of inheritance tree, number of children.
+
+The CCCC static analysis tool is capable of measuring most of the metrics identified.
+The main missing metric is lack of cohesion of methods.
```

Comment:
This sentence is a little confusing to me.  Is "lack of cohesion methods" a metric?

Reply:
Yes, lack of cohesion of methods (LCOM, sometimes LCOM2) is a metric generated by some tools, but none of the open source ones I have found so far.